// @worklouder/device-kit-oai/dist/rpc_api_oai/types/oai_lighting_effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAILightingEffect = void 0;
/**
 * Built-in LED animation effects supported by Work Louder device firmware.
 * Used for the ambient ring, keys backlight, and per-thread lighting on OpenAI
 * integration builds.
 */
var OAILightingEffect;
(function (OAILightingEffect) {
    /** LEDs are turned off. */
    OAILightingEffect[OAILightingEffect["off"] = 0] = "off";
    /** All LEDs show a single static color. */
    OAILightingEffect[OAILightingEffect["solid"] = 1] = "solid";
    /** A colored segment travels across the LED strip. */
    OAILightingEffect[OAILightingEffect["snake"] = 2] = "snake";
    /** Continuously cycles through the full hue spectrum. */
    OAILightingEffect[OAILightingEffect["rainbow"] = 3] = "rainbow";
    /** LEDs fade in and out in a breathing pattern. */
    OAILightingEffect[OAILightingEffect["breath"] = 4] = "breath";
    /** A smooth color gradient spans the LED strip. */
    OAILightingEffect[OAILightingEffect["gradient"] = 5] = "gradient";
    /** LEDs fade in and out in a breathing pattern. Starts at 0.5 brighness and goes to 1 */
    OAILightingEffect[OAILightingEffect["shallowBreath"] = 6] = "shallowBreath";
})(OAILightingEffect || (exports.OAILightingEffect = OAILightingEffect = {}));
