// @worklouder/device-kit-oai/dist/rpc_api_oai/rpc_api_oai.js
"use strict";
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCApiOAI = void 0;
const wl_device_kit_1 = require("@worklouder/wl-device-kit");
const notify_keys_1 = require("./types/notify_keys");
// Vendor specific.
var VendorJsonRpcMethods;
(function (VendorJsonRpcMethods) {
    /** Per-thread accent lighting. */
    VendorJsonRpcMethods["ThreadsLighting"] = "v.oai.thstatus";
    /** Keys and ambient zone lighting. */
    VendorJsonRpcMethods["RgbConfig"] = "v.oai.rgbcfg";
})(VendorJsonRpcMethods || (VendorJsonRpcMethods = {}));
/**
 * Device API class for the OpenAI team: status queries, vendor lighting
 * (threads + keys/ambient config), generic lighting previews, and device
 * notifications.
 */
class RPCApiOAI {
    /**
     * @param deviceComm - The {@link WLDeviceComm} transport to send requests through.
     * @param logger - {@link Logger} instance for diagnostic output.
     */
    constructor(deviceComm, logger) {
        this.log = logger ?? wl_device_kit_1.noopLogger;
        this.deviceComm = deviceComm;
        this.api = new wl_device_kit_1.WLRPCApi(deviceComm, logger);
    }
    /**
     * Subscribes to custom HID key events coming from the device.
     *
     * The callback receives a {@link HIDKey} payload describing the received HID
     * event.
     *
     * @param callback - Invoked whenever a custom HID event is received.
     * @returns Unsubscribe function that removes the notification handler.
     */
    onHidReceived(callback) {
        this.deviceComm.addNotifyHandler(notify_keys_1.NotifyKeys.hidReceived, (params) => {
            const hid = {
                key: params.k,
                act: params.act,
                agent: params.ag,
            };
            callback(hid);
        });
        return () => this.deviceComm.removeNotifyHandler(notify_keys_1.NotifyKeys.hidReceived);
    }
    /**
     * Subscribes to joystick movement notifications.
     *
     * The callback receives a {@link JoystickPos} payload describing the current
     * joystick position.
     *
     * @param callback - Invoked whenever joystick position changes.
     * @returns Unsubscribe function that removes the notification handler.
     */
    onJoystickMove(callback) {
        this.deviceComm.addNotifyHandler(notify_keys_1.NotifyKeys.joystickMove, (params) => {
            const joystickPos = {
                angle: params.a,
                distance: params.d,
            };
            callback(joystickPos);
        });
        return () => this.deviceComm.removeNotifyHandler(notify_keys_1.NotifyKeys.joystickMove);
    }
    /**
     * Queries the firmware version via JSON-RPC.
     * @returns The firmware version string.
     */
    getFirmwareVersion() {
        return this.api.getFirmwareVersion();
    }
    /**
     * Retrieves the current device status snapshot.
     * @returns A {@link WLDeviceStatus} object.
     */
    getDeviceStatus() {
        return this.api.getDeviceStatus();
    }
    /**
     * Sends a real-time lighting preview to the device.
     * @param config - Backlight and underglow configuration to preview.
     */
    sendLightingPreview(config) {
        return this.api.sendLightingPreview(config);
    }
    /**
     * Sends per-thread lighting state for UI threads on the device.
     *
     * Only the thread id is required on each entry. See {@link ThreadLighting}
     * for optional color, brightness, effect, speed, and sync flags (brightness
     * 0 = off through 1 = full on; speed 0 = stopped through 1 = fast).
     *
     * @param threads - One object per thread; omit optional fields to leave
     *   those parameters unchanged on the device.
     * @returns `true` if the RPC succeeds, `false` if the call throws (errors
     *   are logged via the optional {@link Logger}).
     */
    async sendThreadsLighting(threads) {
        try {
            let minimized = threads.map((thread) => {
                return {
                    id: thread.id,
                    c: thread.color,
                    b: thread.brightness,
                    e: thread.effect,
                    s: thread.speed,
                    sk: thread.syncKeysLighting !== undefined ? thread.syncKeysLighting ? 1 : 0 : undefined,
                    sa: thread.syncAmbientLighting !== undefined ? thread.syncAmbientLighting ? 1 : 0 : undefined,
                };
            });
            await this.api.getRpcClient().sendRpcCall({ method: VendorJsonRpcMethods.ThreadsLighting, params: minimized });
            return true;
        }
        catch (error) {
            this.log.error(`|wl_rpc_api| error while sending threads lighting, error: ${error}`);
            return false;
        }
    }
    /**
     * Applies lighting configuration for the outer ambient ring and keys.
     *
     * @param config - {@link OAILightingConfig}
     * @returns `true` if the RPC succeeds, `false` if the call throws (errors
     *   are logged via the optional {@link Logger}).
     */
    async sendLightingConfig(config) {
        try {
            const minimized = {
                ambient: this.minimizeLightingSide(config.ambient),
                keys: this.minimizeLightingSide(config.keys)
            };
            await this.api.getRpcClient().sendRpcCall({ method: VendorJsonRpcMethods.RgbConfig, params: minimized });
            return true;
        }
        catch (error) {
            this.log.error(`|wl_rpc_api| error while sending lighting config, error: ${error}`);
            return false;
        }
    }
    minimizeLightingSide(lightingSide) {
        return {
            e: lightingSide.effect,
            b: lightingSide.brightness,
            s: lightingSide.speed,
            m: lightingSide.magic,
            c: lightingSide.color
        };
    }
}
exports.RPCApiOAI = RPCApiOAI;
exports.default = RPCApiOAI;
