// @worklouder/wl-device-kit/dist/wl_rpc_api/types/lighting_effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightingEffect = void 0;
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
/**
 * Built-in LED animation effects supported by Work Louder device firmware.
 * Used in {@link SideLightingConfig} to control backlight and underglow.
 */
var LightingEffect;
(function (LightingEffect) {
    /** LEDs are turned off. */
    LightingEffect["off"] = "off";
    /** All LEDs show a single static color. */
    LightingEffect["solid"] = "solid";
    /** A colored segment travels across the LED strip. */
    LightingEffect["snake"] = "snake";
    /** Continuously cycles through the full hue spectrum. */
    LightingEffect["rainbow"] = "rainbow";
    /** LEDs fade in and out in a breathing pattern. */
    LightingEffect["breath"] = "breath";
    /** A smooth color gradient spans the LED strip. */
    LightingEffect["gradient"] = "gradient";
})(LightingEffect || (exports.LightingEffect = LightingEffect = {}));
