// @worklouder/wl-device-kit/dist/wl_rpc_api/wl_rpc_api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonRPCMethods = void 0;
const event_emitter_1 = __importDefault(require("../utils/event_emitter"));
const wl_rpc_client_1 = __importDefault(require("../wl_rpc_client/wl_rpc_client"));
const rpc_signal_1 = require("./types/rpc_signal");
const logger_1 = require("../common/types/logger");
/**
 * String constants for every JSON-RPC method supported by Work Louder
 * device firmware. Used internally by {@link WLRPCApi} to construct requests.
 */
var JsonRPCMethods;
(function (JsonRPCMethods) {
    // Legacy (pre-`sys.*`) RPC calls.
    /** Legacy: query firmware version. */
    JsonRPCMethods["VersionLegacy"] = "version";
    /** Legacy: reboot into bootloader. */
    JsonRPCMethods["BootloaderLegacy"] = "bootloader";
    /** Legacy: enter self-test mode. */
    JsonRPCMethods["SelftestLegacy"] = "selftest";
    // System.
    /** Query firmware version (JSON-RPC). */
    JsonRPCMethods["SysVersion"] = "sys.version";
    /** Reboot into bootloader (JSON-RPC). */
    JsonRPCMethods["SysBootloader"] = "sys.bootloader";
    /** Enter self-test / diagnostic mode. */
    JsonRPCMethods["SysSelftest"] = "sys.selftest";
    // Device.
    /** Query the full device status (firmware, profile, layer, battery). */
    JsonRPCMethods["DeviceStatus"] = "device.status";
    // Media playback / metadata.
    /** Send currently playing media metadata to the device display. */
    JsonRPCMethods["MpWriteInfo"] = "mp.write_info";
    /** Stream album artwork to the device display in base64 chunks. */
    JsonRPCMethods["MpWriteArtwork"] = "mp.write_artwork";
    // Filesystem.
    /** List all files on the device filesystem (with optional checksums). */
    JsonRPCMethods["FsList"] = "fs.list";
    /** Read an entire file from the device filesystem (JSON payload). */
    JsonRPCMethods["FsRead"] = "fs.read";
    /** Write an entire file to the device filesystem (JSON payload). */
    JsonRPCMethods["FsWrite"] = "fs.write";
    /** Write binary data to a file in base64-encoded chunks. */
    JsonRPCMethods["FsWriteBin"] = "fs.writebin";
    /** Read binary data from a file in base64-encoded chunks. */
    JsonRPCMethods["FsReadBin"] = "fs.readbin";
    /** Delete a file from the device filesystem. */
    JsonRPCMethods["FsDelete"] = "fs.delete";
    // UI.
    /** Query the currently active screen on the device display. */
    JsonRPCMethods["UiActiveScreen"] = "ui.active_screen";
    /** Set the accent color shown on the device home screen. */
    JsonRPCMethods["UiHomeAccentColor"] = "ui.home_accent_color";
    // Lighting / host.
    /** Send a real-time lighting preview configuration. */
    JsonRPCMethods["LightsPreview"] = "lights.preview";
    /** Notify the device about the currently focused desktop application. */
    JsonRPCMethods["HostFocusedApp"] = "host.focused_app";
})(JsonRPCMethods || (exports.JsonRPCMethods = JsonRPCMethods = {}));
/**
 * High-level API for interacting with Work Louder device firmware over
 * JSON-RPC 2.0 (and legacy `#rpc#` protocol where needed).
 *
 * Built on top of {@link WLDeviceComm} (transport) and {@link WLRPCClient}
 * (request framing), this class exposes typed methods for every firmware
 * capability: system commands, device status, filesystem operations,
 * media playback metadata, lighting control, and UI customisation.
 *
 * Instantiate with an already-created {@link WLDeviceComm} instance
 * (connection does not need to be open yet):
 *
 * ```ts
 * const comm = new WLDeviceCommImpl(logger);
 * const api  = new WLRPCApi(comm, logger);
 *
 * await comm.connect(device);
 * const status = await api.getDeviceStatus();
 * ```
 */
class WLRPCApi {
    /**
     * @param device - The {@link WLDeviceComm} transport to send requests through.
     * @param logger - Optional {@link Logger} instance. If omitted, logging is disabled.
     */
    constructor(device, logger) {
        this.rpcController = new event_emitter_1.default();
        this.deviceComm = device;
        this.rpcClient = new wl_rpc_client_1.default(device, logger);
        this.log = logger ?? logger_1.noopLogger;
    }
    /**
     * Returns the underlying {@link WLRPCClient} used for all requests on this API instance.
     *
     * Prefer the typed methods on {@link WLRPCApi} when available. Use this when you need
     * {@link WLRPCClient.sendRpcCall} or {@link WLRPCClient.sendLegacyRpcRequest} directly
     * (e.g. custom or experimental firmware methods not yet wrapped here).
     *
     * @returns The same client constructed in the constructor; shares transport and logger with this API.
     */
    getRpcClient() {
        return this.rpcClient;
    }
    /**
     * Signals an in-progress or queued RPC operation to abort.
     * Pending chunks for chunked writes are removed from the send queue.
     *
     * @param rpcId - The controller id returned by the `controller` callback
     *   in {@link writeFileChunkedFromStr}.
     */
    abortRpcCall(rpcId) {
        this.log.info("|wl_rpc_api| received abort from rpcId: " + rpcId);
        this.rpcController.emit({ rpcId: rpcId, signal: rpc_signal_1.RpcSignalType.abort });
    }
    /**
     * Queries the firmware version using the legacy `#version#` text protocol.
     * @returns The firmware version string reported by the device.
     */
    async getFirmwareVersionLegacy() {
        return this.rpcClient.sendLegacyRpcRequest(JsonRPCMethods.VersionLegacy);
    }
    /**
     * Queries the firmware version via JSON-RPC (`sys.version`).
     * @returns The firmware version string reported by the device.
     * @throws If the RPC call fails or the device returns an error.
     */
    async getFirmwareVersion() {
        try {
            const response = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.SysVersion });
            return response.result.version;
        }
        catch (error) {
            this.log.error(error);
            throw error;
        }
    }
    /**
     * Retrieves the current device status including firmware version,
     * active profile/layer indices, battery level, and charging state.
     *
     * @returns A validated {@link WLDeviceStatus} snapshot.
     * @throws If the RPC call fails or the response contains invalid data types.
     */
    async getDeviceStatus() {
        try {
            const response = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.DeviceStatus });
            const result = response.result;
            const firmwareVersion = result.version;
            const selectedProfileIndex = result.profile_index;
            const selectedLayerIndex = result.layer_index;
            const batteryPercentage = result.battery;
            const isCharging = result.is_charging;
            if ((firmwareVersion === undefined || typeof firmwareVersion === "string") &&
                (selectedProfileIndex === undefined || typeof selectedProfileIndex === "number") &&
                (selectedLayerIndex === undefined || typeof selectedLayerIndex === "number") &&
                (batteryPercentage === undefined || typeof batteryPercentage === "number") &&
                (isCharging === undefined || typeof isCharging === "boolean")) {
                return {
                    firmwareVersion,
                    selectedProfileIndex,
                    selectedLayerIndex,
                    batteryPercentage,
                    isCharging,
                };
            }
            this.log.info("|wl_rpc_api| data type is invalid");
            throw Error("Data type is invalid");
        }
        catch (error) {
            this.log.error(error);
            throw error;
        }
    }
    /**
     * Reboots the device into bootloader (DFU) mode using the legacy
     * `#bootloader#` text protocol. The device will disconnect after this call.
     */
    async sendIntoBootloaderLegacy() {
        return this.rpcClient.sendLegacyRpcRequest(JsonRPCMethods.BootloaderLegacy);
    }
    /**
     * Reboots the device into bootloader (DFU) mode via JSON-RPC
     * (`sys.bootloader`). The device will disconnect after this call.
     * @throws If the RPC call fails.
     */
    async sendIntoBootloader() {
        try {
            const response = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.SysBootloader });
            return response.result;
        }
        catch (error) {
            this.log.error(error);
            throw error;
        }
    }
    /**
     * Enters self-test / diagnostic mode using the legacy `#selftest#` text protocol.
     */
    async sendIntoSelfTestLegacy() {
        return this.rpcClient.sendLegacyRpcRequest(JsonRPCMethods.SelftestLegacy);
    }
    /**
     * Enters self-test / diagnostic mode via JSON-RPC (`sys.selftest`).
     * @throws If the RPC call fails.
     */
    async sendIntoSelfTest() {
        try {
            const response = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.SysSelftest });
            return response.result;
        }
        catch (error) {
            this.log.error(error);
            throw error;
        }
    }
    /**
     * Sends currently playing media metadata to the device display.
     *
     * @param params - Media playback information.
     * @param params.song_title - Title of the currently playing track.
     * @param params.artist - Artist name.
     * @param params.duration - Total track duration in seconds.
     * @param params.elapsed - Elapsed playback time in seconds.
     * @param params.isPlaying - `true` if playback is active.
     * @returns `true` if the device acknowledged the data, `false` on error.
     */
    async sendMediaInfo({ song_title, artist, duration, elapsed, isPlaying }) {
        try {
            await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.MpWriteInfo, params: { song_title: song_title, artist: artist, elapsed: elapsed, total_duration: duration, is_playing: isPlaying } });
            return true;
        }
        catch (error) {
            this.log.error(error);
            return false;
        }
    }
    /**
     * Lists all files on the device filesystem, including size and checksum.
     *
     * @returns An array of {@link WLDeviceFileSummary} objects, or an empty array on error.
     */
    async getFileList() {
        try {
            const result = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.FsList, params: { checksum: true } });
            const fileList = result.result.map((element) => ({
                name: element.name,
                size: element.size,
                checksum: element.checksum,
            }));
            return fileList;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| Error while getting file list " + error);
            return [];
        }
    }
    /**
     * Reads an entire file from the device filesystem as a JSON payload.
     * Suitable for small files; for large binary files use {@link readFileChunked}.
     *
     * @param fileName - The filename on the device to read.
     * @returns The file contents (parsed JSON), the string `"File does not exist"` if
     *   absent, or `undefined` on other errors.
     */
    async readFile(fileName) {
        try {
            const result = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.FsRead, params: { file: fileName } });
            return result.result;
        }
        catch (error) {
            if (error.message === "File does not exist") {
                return error.message;
            }
            this.log.error(error);
            return;
        }
    }
    /**
     * Writes an entire file to the device filesystem as a JSON payload.
     * Suitable for small files; for large binary files use {@link writeFileChunked}.
     *
     * @param fileName - The target filename on the device.
     * @param data - The file contents to write (will be serialized as the `data` field).
     * @returns The device's acknowledgment result, or `undefined` on error.
     */
    async writeFile(fileName, data) {
        try {
            const result = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.FsWrite, params: { file: fileName, data: data } });
            return result.result;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| Error while writing file " + error);
            return;
        }
    }
    /**
     * Writes a file to the device filesystem by encoding a string into base64
     * chunks of 4096 bytes each, sent sequentially via `fs.writebin`.
     *
     * Supports cooperative cancellation: the `controller` callback receives a
     * UUID that can later be passed to {@link abortRpcCall} to cancel
     * remaining chunks.
     *
     * @param fileName - The target filename on the device.
     * @param inputString - The UTF-8 string data to write.
     * @param controller - Called immediately with a cancellation id (`rpcId`).
     *   Store this id and pass it to {@link abortRpcCall} to cancel the write.
     * @returns `true` if the entire file was written successfully, `false` on error or abort.
     */
    async writeFileChunkedFromStr(fileName, inputString, controller) {
        const uuid = crypto.randomUUID();
        controller(uuid);
        let chunkedRpcCalls = [];
        let operationCompleted = false;
        const unsubController = this.rpcController.on((rpcSignal) => {
            if (operationCompleted || rpcSignal.rpcId !== uuid || rpcSignal.signal !== rpc_signal_1.RpcSignalType.abort) {
                return;
            }
            operationCompleted = true;
            unsubController?.();
            chunkedRpcCalls.forEach((promise) => {
                this.log.info(`|wl_rpc_api| aborted rpc call id: ${promise.id}`);
                this.deviceComm.abortJsonRpcRequest(promise.id.toString());
            });
            return false;
        });
        try {
            const encoder = new TextEncoder();
            const data = encoder.encode(inputString);
            const chunkSize = 4096;
            const chunks = this.splitIntoBase64Chunks(data, chunkSize);
            // We send all the chunks so that they can get put into queue
            chunks.forEach((chunk, index) => {
                const id = this.rpcClient.createRpcId();
                chunkedRpcCalls.push({
                    id: id,
                    promise: this.rpcClient.sendRpcCall({
                        method: JsonRPCMethods.FsWriteBin,
                        params: {
                            file: fileName,
                            data: chunk,
                            append: true,
                            completed: index === chunks.length - 1,
                            offset: chunkSize * index
                        },
                        id: id
                    })
                });
            });
            // Then we await them for the result
            for (let index = 0; index < chunkedRpcCalls.length; index++) {
                // If we abort we get out immediatly
                if (operationCompleted) {
                    break;
                }
                const rpcCall = chunkedRpcCalls[index];
                const result = await rpcCall.promise;
                this.log.debug(result);
            }
            operationCompleted = true;
            unsubController?.();
            return true;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| error while writing chunked data " + error);
            return false;
        }
        finally {
            operationCompleted = true;
            unsubController?.();
        }
    }
    /**
     * Writes binary data to a file on the device filesystem in base64-encoded
     * chunks of 4096 bytes each via `fs.writebin`. All chunks are enqueued
     * upfront and awaited sequentially.
     *
     * @param fileName - The target filename on the device.
     * @param data - The raw binary data to write.
     * @returns `true` on success, `false` on error, or `undefined` on unexpected failure.
     */
    async writeFileChunked(fileName, data) {
        try {
            const chunkSize = 4096;
            let chunks = this.splitIntoBase64Chunks(data, chunkSize);
            // We send all the chunks so that they can get put into queue
            const promises = chunks.map((chunk, index) => this.rpcClient.sendRpcCall({
                method: JsonRPCMethods.FsWriteBin,
                params: {
                    file: fileName,
                    data: chunk,
                    append: true,
                    completed: index === chunks.length - 1,
                    offset: chunkSize * index,
                }
            }));
            // Then we await them for the result
            for (let index = 0; index < promises.length; index++) {
                const promise = promises[index];
                const result = await promise;
                this.log.debug(result);
            }
            return true;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| error while writing chunked data " + error);
            return false;
        }
    }
    /**
     * Reads a binary file from the device filesystem in base64-encoded chunks
     * of up to 4096 bytes via `fs.readbin`, then decodes and concatenates the
     * result into a single string.
     *
     * @param fileName - The filename on the device to read.
     * @returns The decoded file contents as a string, or `undefined` on error.
     */
    async readFileChunked(fileName) {
        try {
            var offset = 0;
            var fileSize = 0;
            var result = "";
            var chunkSize = 4096;
            do {
                const chunk = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.FsReadBin, params: { file: fileName, offset: offset, len: chunkSize } });
                this.log.info("|wl_rpc_api| read chunk response");
                fileSize = chunk.result.total_size;
                offset += chunkSize;
                const remaining = fileSize - offset;
                if ((remaining) < chunkSize) {
                    chunkSize = remaining;
                }
                result += atob(chunk.result.data);
                await new Promise((resolve, reject) => setTimeout(resolve, 50));
            } while (offset < fileSize);
            return result;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| Error while reading chuncked file " + error);
            return;
        }
    }
    /**
     * Deletes a file from the device filesystem.
     *
     * @param fileName - The filename on the device to delete.
     * @returns `true` if the file was deleted successfully, or `undefined` on error.
     */
    async deleteFile(fileName) {
        try {
            await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.FsDelete, params: { file: fileName } });
            return true;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| Error while trying to delete file " + error);
            return;
        }
    }
    /**
     * Streams album artwork image data to the device display in base64-encoded
     * chunks of 4096 bytes via `mp.write_artwork`. Each chunk includes its
     * byte offset and the total image size so the device can reassemble.
     *
     * @param data - Raw image data (e.g. JPEG or PNG bytes).
     * @returns `true` if all chunks were sent successfully, `false` on error.
     */
    async writeMediaArtWork(data) {
        try {
            const chunkSize = 4096;
            const chunks = this.splitIntoBase64Chunks(data, chunkSize);
            for (let index = 0; index < chunks.length; index++) {
                const chunk = chunks[index];
                const response = await this.rpcClient.sendRpcCall({
                    method: JsonRPCMethods.MpWriteArtwork,
                    params: {
                        data: chunk,
                        offset: index * chunkSize,
                        size: data.length
                    }
                });
                if (!response) {
                    return false;
                }
                await new Promise((resolve) => setTimeout(resolve, 50));
            }
            return true;
        }
        catch (error) {
            this.log.error("|wl_rpc_api| error while writing chunked data " + error);
            return false;
        }
    }
    /**
     * Queries the name of the screen currently displayed on the device.
     *
     * @returns The screen name string, or `undefined` on error.
     */
    async getDeviceCurrentScreen() {
        try {
            const response = await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.UiActiveScreen });
            return response.result["screen_name"];
        }
        catch (error) {
            this.log.error(error);
            return undefined;
        }
    }
    /**
     * Sets the accent color displayed on the device's home screen.
     *
     * @param color - A color string (e.g. hex `"#FF5500"` or named color)
     *   accepted by the device firmware.
     */
    async sendHomeAccentColor(color) {
        try {
            await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.UiHomeAccentColor, params: { color: color } });
        }
        catch (error) {
            this.log.error(error);
            return;
        }
    }
    /**
     * Sends a real-time lighting preview to the device. The preview is
     * applied immediately but is **not** persisted to flash storage.
     *
     * @param config - Backlight and underglow configuration to preview.
     */
    async sendLightingPreview(config) {
        try {
            await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.LightsPreview, params: config });
        }
        catch (error) {
            this.log.error(error);
            return;
        }
    }
    /**
     * Sends a message to the custom widget.
     *
     * @param method - The widget/API method name without the `wlsdk.` prefix.
     * @param params - Optional positional parameters forwarded to the device.
     */
    async sendCustomWidgetMessage(method, params) {
        try {
            await this.rpcClient.sendRpcCall({ method: `wlsdk.${method}`, params: params });
        }
        catch (error) {
            this.log.error(error);
            return;
        }
    }
    /**
     * Notifies the device about the currently focused desktop application
     * so it can display context-aware UI or switch profiles.
     *
     * @param app - Information about the focused application.
     */
    async sendFocusApp(app) {
        try {
            await this.rpcClient.sendRpcCall({ method: JsonRPCMethods.HostFocusedApp, params: app });
        }
        catch (error) {
            this.log.error(error);
            return;
        }
    }
    /**
     * Splits a `Uint8Array` into base64-encoded strings, each representing
     * at most `chunkSize` raw bytes.
     *
     * @param data - The raw binary data to split.
     * @param chunkSize - Maximum number of raw bytes per chunk.
     * @returns An array of base64-encoded chunk strings.
     */
    splitIntoBase64Chunks(data, chunkSize) {
        let offset = 0;
        let chunks = [];
        while (offset < data.length) {
            var end = offset + chunkSize;
            if (end >= data.length) {
                end = data.length;
            }
            const subArray = data.subarray(offset, end);
            chunks.push(Buffer
                .from(subArray)
                .toString("base64"));
            offset = end;
        }
        return chunks;
    }
}
exports.default = WLRPCApi;
