// @worklouder/wl-device-kit/dist/wl_rpc_client/wl_rpc_client.js
"use strict";
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
Object.defineProperty(exports, "__esModule", { value: true });
const node_crypto_1 = require("node:crypto");
const string_utils_1 = require("../utils/string_utils");
const logger_1 = require("../common/types/logger");
/**
 * Low-level JSON-RPC 2.0 client that serializes requests, generates
 * randomized call IDs, escapes non-ASCII characters, and delegates
 * transport to a {@link WLDeviceComm} instance.
 *
 * Used internally by {@link WLRPCApi}; consumers should prefer the
 * higher-level API unless building custom RPC flows.
 */
class WLRPCClient {
    /**
     * @param device - The {@link WLDeviceComm} transport to send requests through.
     * @param logger - Optional {@link Logger} instance. If omitted, logging is disabled.
     */
    constructor(device, logger) {
        this.device = device;
        this.log = logger ?? logger_1.noopLogger;
    }
    /**
     * Forwards a legacy `#rpc#args#` request to the device transport.
     *
     * @param rpc - The legacy RPC command name.
     * @param args - Optional argument string.
     * @returns The parsed response payload.
     */
    async sendLegacyRpcRequest(rpc, args) {
        return this.device.sendLegacyRpcRequest(rpc, args);
    }
    /**
     * Constructs a {@link JsonRpcRequest}, serializes it to JSON with
     * Unicode-escaped non-ASCII characters, sends it through the device
     * transport, and parses the JSON-RPC response.
     *
     * Uses a random call ID unless one is explicitly provided via `id`.
     *
     * @param method - The JSON-RPC method name (e.g. `"sys.version"`).
     * @param params - Optional parameters object for the method. Defaults to `null`.
     * @param id - Optional call ID. When omitted a random ID is generated
     *   via {@link createRpcId}.
     * @returns The parsed {@link JsonRpcResponse} on success.
     * @throws {Error} If the device responds with a {@link JsonRpcError}.
     */
    async sendRpcCall({ method, params = null, id }) {
        var idCall = id ?? this.createRpcId();
        const jsonRpcMessage = this.createJsonRpcRequest(method, params, idCall);
        try {
            const jsonRpc = JSON.stringify(jsonRpcMessage);
            const escaped = (0, string_utils_1.escapeUnicode)(jsonRpc);
            this.log.info(`|wl_rpc_client| sending rpc call  id: ${idCall} message: `);
            this.log.debug(escaped);
            const result = await this.device.sendJsonRpcRequest(escaped, idCall.toString());
            const obj = JSON.parse(result);
            if (obj.error !== undefined) {
                const response = obj;
                throw Error(response.error.message);
            }
            this.log.info(`|wl_rpc_client| received answer, id: ${idCall} method: ${method}, message:`);
            this.log.info(result);
            const jsonRpcResponse = obj;
            return jsonRpcResponse;
        }
        catch (error) {
            this.log.error(`|wl_rpc_client| there has been an error while calling the rpc, id: ${idCall}, method: ${method} error: ${error}`);
            throw error;
        }
    }
    /**
     * Generates a random integer (0–998) to use as a JSON-RPC call ID.
     * Useful for pre-generating an ID when the caller needs to track or
     * abort an in-flight request (e.g. chunked writes).
     */
    createRpcId() {
        return (0, node_crypto_1.randomInt)(999);
    }
    /**
     * Builds a {@link JsonRpcRequest} object from the given method, params, and id.
     */
    createJsonRpcRequest(method, params = null, id) {
        return {
            method,
            params,
            id,
        };
    }
}
exports.default = WLRPCClient;
