// @worklouder/device-kit-oai/dist/index.js
"use strict";
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
/**
 * `device-kit-oai` — Work Louder device kit for OpenAI integration.
 *

 * @example
 * ```ts
 * import {
 *   WLDeviceDiscovery,
 *   WLDeviceCommImpl,
 *   RPCApiOAI,
 * } from "@worklouder/device-kit-oai";
 *
 * const discovery = new WLDeviceDiscovery();
 * const [device] = discovery.findWLDevices();
 *
 * const comm = new WLDeviceCommImpl();
 * await comm.connect(device);
 *
 * const api = new RPCApiOAI(comm);
 * const status = await api.getDeviceStatus();
 * console.log(status.firmwareVersion);
 *
 * await comm.disconnect();
 * ```
 *
 * @packageDocumentation
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightingEffect = exports.DeviceLayoutType = exports.DeviceType = exports.ConnectionType = exports.ConnectionEventType = exports.WLRPCClient = exports.WLDeviceCommImpl = exports.WLDeviceDiscovery = void 0;
// Curated RPC API for this integration
__exportStar(require("./rpc_api_oai"), exports);
// Re-export device discovery, communication, and RPC client from the base kit
var wl_device_kit_1 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "WLDeviceDiscovery", { enumerable: true, get: function () { return wl_device_kit_1.WLDeviceDiscovery; } });
var wl_device_kit_2 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "WLDeviceCommImpl", { enumerable: true, get: function () { return wl_device_kit_2.WLDeviceCommImpl; } });
var wl_device_kit_3 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "WLRPCClient", { enumerable: true, get: function () { return wl_device_kit_3.WLRPCClient; } });
var wl_device_kit_4 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "ConnectionEventType", { enumerable: true, get: function () { return wl_device_kit_4.ConnectionEventType; } });
var wl_device_kit_5 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "ConnectionType", { enumerable: true, get: function () { return wl_device_kit_5.ConnectionType; } });
var wl_device_kit_6 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "DeviceType", { enumerable: true, get: function () { return wl_device_kit_6.DeviceType; } });
var wl_device_kit_7 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "DeviceLayoutType", { enumerable: true, get: function () { return wl_device_kit_7.DeviceLayoutType; } });
var wl_device_kit_8 = require("@worklouder/wl-device-kit");
Object.defineProperty(exports, "LightingEffect", { enumerable: true, get: function () { return wl_device_kit_8.LightingEffect; } });
