// @worklouder/wl-device-kit/dist/wl_device_comm/wl_device_comm_impl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
const node_hid_1 = require("node-hid");
const serialport_1 = require("serialport");
const event_emitter_1 = __importDefault(require("../utils/event_emitter"));
const connection_type_1 = require("../common/types/connection_type");
const connection_event_type_1 = require("./types/connection_event_type");
const logger_1 = require("../common/types/logger");
/**
 * Concrete implementation of {@link WLDeviceComm} that communicates with
 * Work Louder devices over USB HID or serial (UART).
 *
 * All outgoing requests are serialized through an internal FIFO queue with a
 * 50 ms inter-request cooldown so the device firmware has time to process each
 * command before the next one arrives.
 *
 * HID packets follow a custom framing protocol:
 * - Byte 0: `0x06` (report ID)
 * - Byte 1: channel (`1` = debug log, `2` = RPC)
 * - Byte 2: payload length
 * - Bytes 3–63: payload (up to 61 bytes per packet; longer messages are split)
 *
 * @see {@link WLDeviceComm} for the interface contract.
 */
class WLDeviceCommImpl {
    /**
     * @param logger - Optional {@link Logger} instance. If omitted, logging is disabled.
     */
    constructor(logger) {
        this.connectionListener = new event_emitter_1.default();
        this.CHANNEL_DEBUG = 1;
        this.CHANNEL_RPC = 2;
        this.connectedDevice = null;
        this.port = null;
        this._responseResolvers = {};
        this._rpcResolver = {};
        this._notifyResolvers = {};
        this.rpcResponse = "";
        this.connectionType = undefined;
        this.queue = [];
        this.pending = false;
        this.buffers = {
            [this.CHANNEL_DEBUG]: '',
            [this.CHANNEL_RPC]: ''
        };
        this.log = logger ?? logger_1.noopLogger;
    }
    /** @inheritDoc */
    onConnectionEvent(callback) {
        return this.connectionListener.on(callback);
    }
    /** @inheritDoc */
    async connect(device) {
        if (device.connectionType === connection_type_1.ConnectionType.serial) {
            this.log.info("|wl_device_comm| connecting with serial");
            return this.connectWithSerial(device.portPath);
        }
        else {
            this.log.info("|wl_device_comm| connecting with hid");
            return this.connectWithHID(device.portPath);
        }
    }
    /**
     * Opens a serial port at the given path (115200 baud) and wires up
     * data/error/close handlers.
     */
    async connectWithSerial(portPath) {
        if (this.port !== null) {
            this.log.info("|wl_device_comm| A device is already connected ");
            return Promise.reject("A device is already connected");
        }
        let resolveCb = null;
        let promise = new Promise((resolve, reject) => {
            resolveCb = resolve;
        });
        let baudRate = 115200;
        const port = new serialport_1.SerialPort({ path: portPath, baudRate: baudRate, autoOpen: false });
        port.on('open', () => { this.log.info("|wl_device_comm| Connection opened"); });
        port.on('close', () => {
            this.queue = [];
            this.connectionType = undefined;
            this.port = null;
            this.log.info("|wl_device_comm| Connection closed");
            this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.DISCONNECTED });
        });
        port.on('error', (err) => {
            this.log.error(err);
            this.queue = [];
            this.connectionType = undefined;
            this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.ERROR, error: err });
        });
        port.open((err) => {
            if (err) {
                this.queue = [];
                this.connectionType = undefined;
                this.port = null;
                this.log.error(err);
                this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.ERROR, error: err });
                resolveCb(false);
            }
            else {
                this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.CONNECTED });
                resolveCb(true);
            }
        });
        port.on('data', this.parseSerialRpcData.bind(this));
        const parser = port.pipe(new serialport_1.DelimiterParser({ delimiter: '\n' }));
        parser.on('data', this.parseSerialData.bind(this)); // emits data after every '\n'
        this.connectionType = connection_type_1.ConnectionType.serial;
        this.port = port;
        return promise;
    }
    /**
     * Opens a HID device at the given path and wires up data/error/close
     * handlers. On macOS the device is opened in non-exclusive mode.
     */
    async connectWithHID(portPath) {
        if (this.connectedDevice !== null) {
            this.log.info("|wl_device_comm| A device is already connected ");
            return Promise.reject("A device is already connected");
        }
        try {
            let connectingDevice = process.platform === "darwin" ? await node_hid_1.HIDAsync.open(portPath, { nonExclusive: true }) : await node_hid_1.HIDAsync.open(portPath);
            connectingDevice.on('close', () => {
                this.cleanOnDisconnect();
                this.log.info("|wl_device_comm| Connection closed");
                this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.DISCONNECTED });
            });
            connectingDevice.on('error', (err) => {
                this.log.error(err);
                this.cleanOnDisconnect();
                this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.ERROR, error: err });
            });
            connectingDevice.on('data', this.parseHIDdata.bind(this));
            this.connectionType = connection_type_1.ConnectionType.hid;
            this.connectedDevice = connectingDevice;
            this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.CONNECTED });
            return true;
        }
        catch (error) {
            this.cleanOnDisconnect();
            this.log.error(error);
            this.connectionListener.emit({ type: connection_event_type_1.ConnectionEventType.ERROR, error: error });
            throw error;
        }
    }
    rejectPendingResolvers() {
        const disconnectError = `{"error": {"message":"device disconnected"}}`;
        for (const key of Object.keys(this._rpcResolver)) {
            if (typeof this._rpcResolver[key] === "function") {
                try {
                    this._rpcResolver[key](disconnectError);
                }
                catch { }
            }
        }
        for (const key of Object.keys(this._responseResolvers)) {
            if (typeof this._responseResolvers[key] === "function") {
                try {
                    this._responseResolvers[key](null);
                }
                catch { }
            }
        }
    }
    cleanOnDisconnect() {
        this.rejectPendingResolvers();
        this.queue = [];
        this.connectionType = undefined;
        this.connectedDevice = null;
        this.port = null;
        this._responseResolvers = {};
        this._rpcResolver = {};
        this.rpcResponse = "";
        this.pending = false;
    }
    /** @inheritDoc */
    cleanCommQueue() {
        this.rejectPendingResolvers();
        this._responseResolvers = {};
        this._rpcResolver = {};
        this.rpcResponse = "";
        this.queue = [];
        this.pending = false;
    }
    /** @inheritDoc */
    async disconnect() {
        if (this.connectionType === connection_type_1.ConnectionType.hid) {
            this.log.info("|wl_device_comm| Disconnecting hid device");
            try {
                await this.connectedDevice?.close();
                this.connectedDevice?.removeAllListeners();
            }
            catch (error) {
                this.log.error("|wl_device_comm| Failed to disconnect");
            }
        }
        else {
            this.log.info("|wl_device_comm| Disconnecting serial device");
            this.port?.close();
        }
        this.cleanOnDisconnect();
    }
    /** @inheritDoc */
    isConnected() {
        if (this.connectionType === connection_type_1.ConnectionType.hid) {
            if (this.connectedDevice !== undefined && this.connectedDevice !== null) {
                return true;
            }
        }
        else if (this.connectionType === connection_type_1.ConnectionType.serial) {
            if (this.port !== undefined && this.port !== null) {
                return true;
            }
        }
        return false;
    }
    /** @inheritDoc */
    async sendLegacyRpcRequest(rpc, args = null) {
        return this.enqueue(() => this._sendLegacyRpcRequest(rpc, args));
    }
    async _sendLegacyRpcRequest(rpc, args = null) {
        if (args === null || args === undefined || args === "") {
            this.sendData(`#${rpc}#\r\n`);
        }
        else {
            this.sendData(`#${rpc}#${args}#\r\n`);
        }
        return new Promise((resolve, reject) => {
            let timer = setTimeout(() => {
                if (this._responseResolvers[rpc]) {
                    reject("Timeout");
                    this._responseResolvers[rpc] = null;
                }
            }, 10000);
            this._responseResolvers[rpc] = (args) => {
                clearTimeout(timer);
                resolve(args);
            };
        });
    }
    /** @inheritDoc */
    sendJsonRpcRequest(request, id) {
        return this.enqueue(() => this._sendJsonRpcRequest(request, id), id);
    }
    _sendJsonRpcRequest(request, id) {
        this.log.debug(`|wl_device_comm| started sending rpc call with id: ${id}`);
        this.sendData(request);
        this.rpcResponse = "";
        return new Promise((resolve, reject) => {
            let timer = setTimeout(() => {
                if (this._rpcResolver[id]) {
                    this._rpcResolver[id] = null;
                }
                reject("Timeout");
            }, 10000);
            this._rpcResolver[id] = (args) => {
                clearTimeout(timer);
                resolve(args);
                this._rpcResolver[id] = null;
            };
        });
    }
    /**
     * Dispatches raw data to the device over the active transport.
     * @param message - The string payload to send.
     */
    async sendData(message) {
        if (this.connectionType === connection_type_1.ConnectionType.serial) {
            return this.sendDataSerial(message);
        }
        return this.sendDataHID(message);
    }
    /**
     * Writes a string directly to the serial port and drains the buffer.
     */
    sendDataSerial(data) {
        this.port?.write(data);
        this.port?.drain();
    }
    /**
     * Frames and sends a string payload over HID. Messages longer than 61 bytes
     * are automatically split into multiple 64-byte HID reports.
     *
     * @returns `true` on success, `false` if the write failed.
     */
    async sendDataHID(message) {
        const max_packet_size = 61;
        // Start timing
        const startTime = Date.now();
        try {
            // If the message is longer than the maximum packet size, split it
            if (message.length > max_packet_size) {
                // log.info(`Message longer than ${max_packet_size} bytes, splitting into multiple packets...`);
                // Convert to buffer once for efficiency
                const messageBuffer = Buffer.from(message);
                // Send the message in chunks of 61 bytes
                let offset = 0;
                let packetNum = 1;
                while (offset < message.length) {
                    // Calculate chunk size (max 61 bytes)
                    const chunkSize = Math.min(max_packet_size, message.length - offset);
                    this.log.debug(`|wl_device_comm| Sending packet ${packetNum} with ${chunkSize} bytes, offset: ${offset}`);
                    // Prepare packet
                    const data = Buffer.alloc(64); // Standard HID report size
                    data[0] = 0x06; // HID byte
                    data[1] = this.CHANNEL_RPC;
                    data[2] = chunkSize; // Length of this chunk
                    // Copy chunk of data
                    messageBuffer.copy(data, 3, offset, offset + chunkSize);
                    // Send the chunk
                    const bytesWritten = await this.connectedDevice?.write(data);
                    // log.info(`Sent packet ${packetNum}: ${bytesWritten} bytes`);
                    // Move to next chunk
                    offset += chunkSize;
                    packetNum++;
                    // Add a small delay between packets if sending multiple
                    if (offset < message.length) {
                        // Wait 10ms between packets
                        //const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
                        //await delay(1);
                    }
                }
                // Calculate and show elapsed time
                const elapsedTime = (Date.now() - startTime) / 1000;
                this.log.debug(`|wl_device_comm| Send complete: ${packetNum - 1} packets, ${message.length} bytes in ${elapsedTime.toFixed(3)} seconds`);
                return true;
            }
            else {
                // For messages that fit in a single packet, use the original approach
                const data = Buffer.alloc(64); // Standard HID report size
                data[0] = 0x06; // HID byte
                data[1] = this.CHANNEL_RPC;
                data[2] = message.length; // Actual length
                Buffer.from(message).copy(data, 3, 0, message.length);
                // Send the message
                const bytesWritten = await this.connectedDevice?.write(data); // Calculate and show elapsed time
                const elapsedTime = (Date.now() - startTime) / 1000;
                this.log.debug(`|wl_device_comm| Sent ${bytesWritten} bytes in ${elapsedTime.toFixed(3)} seconds`);
                return true;
            }
        }
        catch (error) {
            this.log.error(`|wl_device_comm| Error sending message: ${error.message}`);
            this.log.error(error.stack);
            return false;
        }
    }
    /**
     * Processes an incoming HID report, demultiplexes by channel (debug vs RPC),
     * and forwards complete newline-terminated lines to the appropriate parser.
     */
    parseHIDdata(data) {
        // Process according to protocol
        try {
            const packet = this.parseHIDReport(data);
            const channel = packet.channel;
            const textData = packet.payload;
            // Add data to the appropriate buffer
            if (this.buffers[channel] === undefined) {
                this.buffers[channel] = '';
            }
            // Look for line endings in the received data
            this.buffers[channel] += textData;
            const lines = this.buffers[channel].split(/\r?\n/);
            // We only get here if we have a complete message or we have more than one message
            if (lines.length > 1 || textData.endsWith('\n') || textData.endsWith('\r')) {
                // All complete lines should be printed
                for (let i = 0; i < lines.length - 1; i++) {
                    const line = lines[i].trim();
                    if (line) {
                        const prefix = channel === this.CHANNEL_DEBUG ? '[LOG]' : '[RPC]';
                        if (channel === this.CHANNEL_RPC) {
                            this.log.debug(`|wl_device_comm| ${prefix} ${line}`);
                            try {
                                this.parseRpcData(line);
                            }
                            catch (error) {
                                this.log.error(`|wl_device_comm| failed to parse: ${error.message}`);
                            }
                        }
                        else {
                            this.log.info(`|wl_device_comm| ${prefix} ${line}`);
                        }
                    }
                }
                // We keep the incomplete line, if all are completed we have empty string here
                this.buffers[channel] = lines[lines.length - 1];
            }
        }
        catch (error) {
            this.log.error(`|wl_device_comm| [ERROR] Failed to parse packet: ${error.message}`);
            this.buffers[this.CHANNEL_DEBUG] = "";
            this.buffers[this.CHANNEL_RPC] = "";
        }
    }
    /**
     * Extracts channel, length, and UTF-8 payload from a raw 64-byte HID report.
     */
    parseHIDReport(data) {
        // Skip first byte (HID stuff)
        const channel = data[1];
        const length = data[2];
        const payload = data.slice(3, 3 + length);
        return {
            channel,
            length,
            payload: Buffer.from(payload).toString('utf8')
        };
    }
    /** Decodes a serial data buffer and routes it to the legacy response parser. */
    parseSerialData(data) {
        const decoded = new TextDecoder().decode(data);
        this.parseData(decoded);
    }
    /**
     * Parses legacy `#rpc#response#` formatted data received over serial and
     * resolves the matching pending request promise.
     */
    parseData(data) {
        let str = data.replace(/[\r\n]+$/, "");
        try {
            if (this.checkMessage(str, "version")) {
                let version = this.extractResponse(str);
                this._responseResolvers["version"](version);
                this._responseResolvers["version"] = null;
            }
            else if (this.checkMessage(str, "dfu")) {
                let response = this.extractResponse(str);
                if (response === "ok")
                    this._responseResolvers["dfu"](true);
                this._responseResolvers["dfu"] = null;
            }
            else if (this.checkMessage(str, "selftest")) {
                let response = this.extractResponse(str);
                if (response === "ok")
                    this._responseResolvers["selftest"](true);
                this._responseResolvers["selftest"] = null;
            }
            else if (this.checkMessage(str, "bootloader")) {
                let response = this.extractResponse(str);
                if (response === "ok")
                    this._responseResolvers["bootloader"](true);
                this._responseResolvers["bootloader"] = null;
            }
        }
        catch (error) {
            this.log.error(error);
        }
    }
    /** Decodes a serial data buffer and routes it to the JSON-RPC response parser. */
    parseSerialRpcData(data) {
        const string = new TextDecoder().decode(data);
        this.parseRpcData(string);
    }
    /**
     * Accumulates incoming JSON fragments and, once a complete JSON object is
     * received, either resolves a pending RPC promise or dispatches a device
     * notification via {@link handleNotifyMessage}.
     *
     * @returns `true` if the data was consumed or ignored; `false` if more data is needed.
     */
    parseRpcData(data) {
        const string = data;
        // If we have no accumulated response, check if this chunk contains a JSON start
        if (this.rpcResponse.length === 0) {
            const jsonStart = string.indexOf('{');
            if (jsonStart === -1) {
                return true; // No JSON start found, ignore this chunk
            }
            // Only add from the first opening brace
            this.rpcResponse = string.slice(jsonStart);
        }
        else {
            // We're already accumulating JSON, append the entire chunk
            this.rpcResponse += string;
        }
        try {
            const parsedData = JSON.parse(this.rpcResponse);
            // All messages should at have either id or method
            if (!parsedData.id && !parsedData.i && !parsedData.method && !parsedData.m) {
                this.log.warn(`|wl_device_comm| received rpc call without id and method`);
                this.rpcResponse = "";
            }
            if ((!parsedData.id && parsedData.method) || (!parsedData.i && parsedData.m)) {
                this.handleNotifyMessage(parsedData);
                this.rpcResponse = "";
            }
            else {
                if (this._rpcResolver[parsedData.id]) {
                    this._rpcResolver[parsedData.id](this.rpcResponse);
                    this._rpcResolver[parsedData.id] = null;
                    this.rpcResponse = "";
                }
                else {
                    this.log.warn(`|wl_device_comm| No resolver found for id: ${parsedData.id}, response: ${this.rpcResponse}`);
                }
            }
        }
        catch (error) {
            // log.error(error);
            // If parsing fails, we need more data
            return false;
        }
        return true;
    }
    /** @inheritDoc */
    addNotifyHandler(method, handler) {
        this.log.info("|wl_device_comm| add notify called");
        this._notifyResolvers[method] = handler;
    }
    /** @inheritDoc */
    removeNotifyHandler(method) {
        this._notifyResolvers[method] = undefined;
    }
    /**
     * Tests whether `msg` matches the legacy response pattern `#rpc#...#`.
     */
    checkMessage(msg, rpc) {
        let re = new RegExp(String.raw `\#${rpc}#.*#$`, "g");
        return msg.search(re) !== -1;
    }
    /**
     * Extracts the payload between the second and third `#` delimiters in a
     * legacy `#rpc#payload#` response string.
     */
    extractResponse(msg) {
        let re = new RegExp(String.raw `\#.*?\#(.*?)\#$`, "g");
        let match = re.exec(msg);
        if (match && match[1]) {
            return match[1];
        }
        return "";
    }
    /**
     * Dispatches a device-initiated notification to the registered handler
     * for its `method` (or compact `m`) field.
     */
    handleNotifyMessage(notifyMessage) {
        const method = notifyMessage.method || notifyMessage.m;
        if (this._notifyResolvers[method] === undefined) {
            this.log.debug(`|wl_device_comm| cannot find notify with method ${method}`);
            return false;
        }
        this._notifyResolvers[method](notifyMessage.params || notifyMessage.p);
        return true;
    }
    /**
     * Adds an async task to the sequential send queue. Tasks execute one at a
     * time with a 50 ms cooldown between each to avoid overwhelming the device.
     *
     * @param fn - The async function to execute when this task reaches the front.
     * @param id - Optional identifier used by {@link abortJsonRpcRequest} to remove the task before execution.
     * @returns A promise that resolves/rejects with the result of `fn`.
     */
    enqueue(fn, id) {
        return new Promise((resolve, reject) => {
            // Add the task to the queue
            this.queue.push({ id, fn, resolve, reject });
            this.log.info("|wl_device_comm| Added call to queue, new queue length: " + this.queue.length);
            // Kick off processing if not already running
            this.process();
        });
    }
    /** @inheritDoc */
    async abortJsonRpcRequest(id) {
        this.queue = this.queue.filter((task) => task.id !== id);
        if (this._rpcResolver[id]) {
            this._rpcResolver[id](`{"error": {"message":"task aborted"}}`);
            this._rpcResolver[id] = null;
        }
    }
    async process() {
        // If a task is already running, do nothing
        if (this.pending) {
            this.log.debug("|wl_device_comm| task already in progress, waiting for it to finish");
            return;
        }
        this.pending = true;
        const task = this.queue.shift();
        if (!task) {
            this.log.debug("|wl_device_comm| no more task to process");
            this.pending = false;
            return;
        }
        try {
            this.log.debug("|wl_device_comm| processing next task");
            const result = await task.fn();
            task.resolve(result);
        }
        catch (err) {
            this.log.debug("|wl_device_comm| task failed");
            task.reject(err);
        }
        finally {
            // We need a small delay between calls so that the device has time to handle the request
            await new Promise((r) => setTimeout(r, 50));
            this.pending = false;
            // Process next task in the queue
            this.process();
        }
    }
}
exports.default = WLDeviceCommImpl;
