// @worklouder/wl-device-kit/dist/utils/event_emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Lightweight, type-safe, single-event emitter.
 *
 * Unlike Node's built-in `EventEmitter`, this class is generic over a single
 * payload type `T` and returns an unsubscribe function from {@link on},
 * making cleanup straightforward and leak-resistant.
 *
 * @typeParam T - The payload type delivered to every listener on {@link emit}.
 */
class EventEmitter {
    constructor() {
        this.listeners = new Set();
        this._maxListeners = 50;
    }
    /**
     * Sets the maximum number of listeners allowed before new registrations
     * are silently dropped (a guard against accidental leaks).
     *
     * @param n - A non-negative finite number. Fractional values are floored.
     * @throws {TypeError} If `n` is negative or not finite.
     */
    setMaxListeners(n) {
        if (!Number.isFinite(n) || n < 0)
            throw new TypeError("maxListeners must be a non-negative finite number");
        this._maxListeners = Math.floor(n);
    }
    /**
     * Registers a listener that is invoked every time {@link emit} is called.
     *
     * @param callback - Function called with the emitted payload.
     * @returns An unsubscribe function. Call it to remove this listener.
     */
    on(callback) {
        if (this.listeners.size >= this._maxListeners) {
            console.error(`EventEmitter: listener limit reached (${this._maxListeners}). Dropping new listener.`);
            return () => { };
        }
        this.listeners.add(callback);
        return () => this.listeners.delete(callback);
    }
    /**
     * Synchronously invokes every registered listener with the given payload.
     * Exceptions thrown by individual listeners are caught and logged so that
     * one failing listener does not prevent others from executing.
     *
     * @param data - The payload to deliver to all listeners.
     */
    emit(data) {
        for (const cb of this.listeners) {
            try {
                cb(data);
            }
            catch (err) {
                console.error("EventEmitter listener threw:", err);
            }
        }
    }
    /** Returns the number of currently registered listeners. */
    listenerCount() {
        return this.listeners.size;
    }
    /** Removes all registered listeners. */
    clear() {
        this.listeners.clear();
    }
}
exports.default = EventEmitter;
