// @worklouder/wl-device-kit/dist/wl_device_discovery/wl_device_discovery.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (C) 2026 Work Louder Inc. - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Cristian Duciuc <cristian@worklouder.cc>, March 2026
 */
const nodeHID = __importStar(require("node-hid"));
const serialport_1 = require("serialport");
const connection_type_1 = require("../common/types/connection_type");
const device_registry_1 = require("./const/device_registry");
const device_type_1 = require("../common/types/device_type");
const device_layout_type_1 = require("../common/types/device_layout_type");
const logger_1 = require("../common/types/logger");
const wl_const_1 = require("./const/wl_const");
/**
 * Enumerates connected Work Louder hardware by scanning USB HID and serial
 * ports. This is the primary entry point for discovering devices before
 * connecting to them via {@link WLDeviceComm}.
 *
 * Typical usage:
 * ```ts
 * const discovery = new WLDeviceDiscovery();
 * const devices = discovery.findWLDevices();
 * const bootloaders = await discovery.findWLBootloaderDevices();
 * ```
 */
class WLDeviceDiscovery {
    /**
     * @param logger - Optional {@link Logger} instance. If omitted, logging is disabled.
     */
    constructor(logger) {
        this.log = logger ?? logger_1.noopLogger;
    }
    /**
     * Synchronously scans all connected USB HID devices and returns those
     * that match known Work Louder Product IDs on the vendor-specific HID
     * usage page (`0xFF00`).
     *
     * @param filter - When provided, only devices whose {@link WLDevice.deviceType}
     *   is in this array are returned. Pass `undefined` or omit to get all devices.
     * @returns An array of discovered {@link WLDevice} descriptors (may be empty).
     */
    findWLDevices(filter) {
        const hidDevices = nodeHID.devices();
        // First try with manufacturer check
        const wlDevices = this.filterWLDevices(hidDevices);
        // If no filter is passed we return everything
        if (!filter) {
            return wlDevices;
        }
        return wlDevices.filter((device) => filter.includes(device.deviceType));
    }
    /**
     * Filters raw HID device descriptors down to recognized Work Louder
     * devices. First attempts matching by manufacturer string + VID; falls
     * back to VID-only if no manufacturer match is found.
     */
    filterWLDevices(hidDevices) {
        // First try with manufacturer check
        let workLouderDevices = hidDevices.filter(device => device.manufacturer !== undefined &&
            wl_const_1.WL_MANUFACTURER.some(manufacturer => device.manufacturer.includes(manufacturer)) &&
            device.vendorId === wl_const_1.WL_VID);
        // If no devices found with manufacturer check, try without manufacturer check
        if (workLouderDevices.length === 0) {
            workLouderDevices = hidDevices.filter(device => device.vendorId === wl_const_1.WL_VID);
        }
        return workLouderDevices.map((device) => {
            if (!device.path || !device.productId || device.usagePage !== 0xFF00) {
                return;
            }
            const wlDevice = device_registry_1.DEVICE_REGISTRY.get(device.productId);
            if (!wlDevice) {
                return;
            }
            const isUsb = (device.release & 0x0003) === 0x00;
            return {
                portPath: device.path,
                devicePid: device.productId.toString(),
                connectionType: connection_type_1.ConnectionType.hid,
                deviceType: wlDevice.type,
                layoutType: wlDevice.layout,
                isUsbConnection: isUsb,
            };
        }).filter((device) => device !== undefined);
    }
    /**
     * Asynchronously scans serial ports for Work Louder devices in bootloader
     * (DFU) mode. Bootloader devices appear as serial ports with the
     * Espressif manufacturer string or (on Windows) a serial number ending
     * in `"0000"`.
     *
     * @returns An array of discovered bootloader {@link WLDevice} descriptors (may be empty).
     */
    async findWLBootloaderDevices() {
        const serialDevices = await serialport_1.SerialPort.list();
        return this.filterBootloaderDevices(serialDevices);
    }
    /**
     * Filters serial port descriptors to find Work Louder bootloader devices.
     * Uses platform-specific heuristics: on Windows, matches serial numbers
     * ending in `"0000"`; on other platforms, matches the `"Espressif"` manufacturer.
     */
    filterBootloaderDevices(serialPorts) {
        const isWin = (process.platform === "win32");
        try {
            let bootloaderDevices = [];
            serialPorts.forEach((port) => {
                let vendorId = parseInt(port.vendorId ?? '', 16);
                if (vendorId !== wl_const_1.WL_VID) {
                    return;
                }
                const productId = parseInt(port.productId, 16).toString();
                const device = {
                    portPath: port.path,
                    devicePid: productId,
                    connectionType: connection_type_1.ConnectionType.serial,
                    deviceType: device_type_1.DeviceType.Bootloader,
                    layoutType: device_layout_type_1.DeviceLayoutType.unknown,
                    isUsbConnection: true,
                };
                if (isWin && port.serialNumber?.endsWith("0000")) {
                    bootloaderDevices.push(device);
                }
                else if (port.manufacturer === "Espressif") {
                    bootloaderDevices.push(device);
                }
            });
            return bootloaderDevices;
        }
        catch (error) {
            this.log.error(error);
            return [];
        }
    }
}
exports.default = WLDeviceDiscovery;
